packages/php-wasm/universal/src/lib/base-php.ts

```diff
@@ -408,11 +408,14 @@ export abstract class BasePHP implements IsomorphicLocalPHP {
 	}
 
 	#setRequestBody(body: string) {
+		const size = this[__private__dont__use].lengthBytesUTF8(body);
+		const addr = this[__private__dont__use]._malloc(size + 1);
+		this[__private__dont__use].stringToUTF8(body, addr, size + 1);
```

Comment:
Won't this allocate more and more heap memory on every request, and eventually cause a crash? Also, what happens when body isn't a UTF8 string?

Reply:
That value is now being freed here:

https://github.com/WordPress/wordpress-playground/blob/4929a9508d883369710f2079b20a5d6bf887f9e6/packages/php-wasm/universal/src/lib/base-php.ts#L420

I'm not entirely sure how it would handle every case when an incorrect type is handled, its the same thing it would do for the stack:

https://github.com/emscripten-core/emscripten/blob/e3062090695bab845398890c510614392fbe4bc0/src/library_ccall.js#L36

https://github.com/emscripten-core/emscripten/blob/e3062090695bab845398890c510614392fbe4bc0/src/library_strings.js#L486-L493

Reply:
> I'm not sure how it would handle every case when an incorrect type is handled,

Could you check what would happen when body consists of non-utf8 bytes?

Reply:
> That value is now being freed here:

If I understand correctly, the request body is copied onto the heap by `_malloc`, and then passed to `wasm_set_request_body` where it's copied on the heap once again. Could `wasm_set_request_body` just accept a heap pointer to avoid allocating 2x the memory?

Reply:
@adamziel I don't mind doing that. I just had each "layer" managing its own memory.

Reply:
https://github.com/WordPress/wordpress-playground/blob/5024842b90d3180155eec25725ba032711c95f80/packages/php-wasm/compile/php/php_wasm.c#L773-L775

Reply:
Let's add a comment here to explain where that heap allocation is freed later on. Let's also add a comment to the C function to explain that `request_body` is expected to be a heap pointer.

Also, not a blocker here, but for posterity – there's a memory leak here if I call the `#setRequestBody` method twice (the previous body will never be freed). That's the same as before, hence why it's not a blocker.

Reply:
We could also free the pointer it in that case to prevent that leak:

https://github.com/WordPress/wordpress-playground/blob/934826891c3fb24efd76ac228f73a5196e424e9d/packages/php-wasm/compile/php/php_wasm.c#L767-L785

We're using a similar pattern for `additional_phpini_entries`:

https://github.com/WordPress/wordpress-playground/blob/934826891c3fb24efd76ac228f73a5196e424e9d/packages/php-wasm/compile/php/php_wasm.c#L555-L560